File: raw_control.py
import serial
import struct
import time
import numpy as np

SPE = 0x06
SMOT = 0x01
SMOD = 0x08

RES = 0x02
MOD = 0x0A
SPEED_DUAL = 0x02
POW = 0x05
XXX = 0xFF

# start, data_length, sequence, comand_type, __, value, value, crc, crc, end
REScommand    = b'/\x00\x00' + struct.pack('b', RES) + struct.pack('b', SMOD) + b'\x00\x00\n';
MODcommand    = ['/', 2,  0, MOD, SMOD,   SPEED_DUAL, 250, 0,   0, '\n'];
PWRcommand    = ['/', 2,  0, POW, SMOT,   0, 0, 		     0,   0, '\n'];
SPEcommand    = ['/', 2,  0, SPE, SMOT,   0, 0, 		     0,   0, '\n'];
XXXcommand    = ['/', 0,  0, XXX, SMOT,         		     0,   0, '\n'];

sequence = 0x00

class Message:
  def __init__(self, command_type, modality = SMOT):
    self.start = b'/'
    self.length = 0
    self.sequence = 0
    self.command_type = command_type
    self.modality = modality
    self.value = 13107 # 0x3333
    self.value = 65535 # 0xffff
    self.end = b'\n'

  def build_buffer(self):
    buff = self.start
    buff += struct.pack('B', self.length)
    buff += struct.pack('B', self.sequence)
    buff += struct.pack('B', self.command_type)
    buff += struct.pack('B', self.modality)
    buff += struct.pack('>H', self.value)
    crc = self.calc_crc(buff)
    print(hex(crc))
    buff += struct.pack('H', crc)
    buff += self.end
    return buff

  def calc_crc(self, buff):
    crc = np.uint16(0)
    count = buff[1] + 5
    for i in range(count):
      crc = crc ^ buff[i] << 8
      for j in range(8):
        if (crc & 0x8000):
          crc = np.uint16(crc << 1) ^ 0x1021
        else:
          crc = crc << 1
    return crc


def send_data(ser, buff):
  global sequence
  dataLength = buff[1]
  # buff[2] = sequence
  crc = calc_crc(buff)
  buff[dataLength + 5] = crc & 0xFF
  buff[dataLength + 6] = (crc >> 8) & 0xFF
  buff[dataLength + 7] = '\n'
  return ser.write(buff)


reset_command = Message(RES, SMOD)
speed_mode = Message(MOD, SMOD)
speed_mode.value = int.from_bytes(
  struct.pack('BB', 0x02, 250), byteorder='big')
speed_command = Message(SPE, SMOT)
speed_command.value = 5000

power_command = Message(POW, SMOT)
power_command.length = 2
power_command.value = 0xAA
power_command.build_buffer()
exit()

if __name__ == '__main__':
  ser = serial.Serial('/dev/ttyUSB0', baudrate=115200)

  reset_command.sequence = sequence
  msg = reset_command.build_buffer()
  print('MSG', msg)
  ser.write(msg)
  #print(ser.readline())
  #exit()
  sequence += 0x10

  speed_mode.sequence = sequence
  msg = speed_mode.build_buffer()
  ser.write(msg)
  sequence += 0x10

  #exit()
  while True:
    speed_command.sequence = sequence
    msg = speed_command.build_buffer()
    print(msg)
    ser.write(msg)
    sequence += 0x10

    if(sequence > 0xff):
      sequence = sequence % 0xff

    #time.sleep(0.1)	
    #print(ser.readline())

File: raw_control.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <termios.h>
#include <stdint.h>

#define SPE 0x06
#define SMOT 0x01
#define SMOD 0x08

#define RES 0x02
#define MOD 0x0A
#define SPEED_DUAL 0x02
#define POW 0x05
#define XXX 0xFF

#define BUFF_SIZE 20

unsigned char sequence = 0x00;

uint8_t	REScommand[]    = {'/', 0,  0, RES, SMOD,         		     0,   0, '\n'};

struct Message {
    unsigned char start;
    unsigned char length;
    unsigned char sequence;
    unsigned char command_type;
    unsigned char modality;
    unsigned short value;
    unsigned short crc;
    unsigned char end;
};

void build_buffer(struct Message* msg, unsigned char* buff);
unsigned short calc_crc(unsigned char* buff, int count);
uint16_t CalcCRC(uint8_t *ptr);
int send_data(int fd, unsigned char* buff, int size);
void print_buffer(unsigned char* buff, int count);

void read_data(int fd){
  char buff[5];

  // Read from the serial port into the buffer
  ssize_t bytesRead = read(fd, buff, sizeof(buff) - 1);

  if (bytesRead == -1) {
      perror("Error reading from serial port");
      // Handle error
  } else {
      // Null-terminate the received data
      buff[bytesRead] = '\0';

      // Process the received data
      printf("Received data: %s\n", buff);
  }
}

int main() {
    int fd;
    struct termios serial;

    fd = open("/dev/ttyUSB0", O_RDWR | O_NOCTTY | O_NDELAY);
    if (fd == -1) {
        perror("Error opening serial port");
        return 1;
    }

    tcgetattr(fd, &serial);
    cfsetispeed(&serial, B115200);
    cfsetospeed(&serial, B115200);

    serial.c_cflag |= (CLOCAL | CREAD);
    serial.c_cflag &= ~PARENB;
    serial.c_cflag &= ~CSTOPB;
    serial.c_cflag &= ~CSIZE;
    serial.c_cflag |= CS8;

    tcsetattr(fd, TCSANOW, &serial);

    struct Message reset_command = {'/', 0, 0, RES, SMOD, 0, 0, '\n'};
    struct Message speed_mode = {'/', 2, 0, MOD, SMOD, 0x02FA, 0, '\n'};
    struct Message speed_command = {'/', 2, 0, SPE, SMOT, 2000, 0, '\n'};

    reset_command.sequence = sequence;
    unsigned char buff[BUFF_SIZE];
    build_buffer(&reset_command, buff);
    write(fd, buff, sizeof(buff)); // Adjust the number of bytes to write
    read_data(fd);

    sequence += 0x10;

    speed_mode.sequence = sequence;
    build_buffer(&speed_mode, buff);
    write(fd, buff, speed_mode.length + 8); // Adjust the number of bytes to write
    sequence += 0x10;

    while (1) {
        speed_command.sequence = sequence;
        build_buffer(&speed_command, buff);
        write(fd, buff, speed_command.length + 8); // Adjust the number of bytes to write
        sequence += 0x10;
        usleep(100000); // Sleep for 100 milliseconds
    }

    close(fd);
    return 0;
}

void print_buffer(unsigned char* buff, int size){
  printf("BUFF: ");
  for(int i = 0; i < size; i++){
    if(buff[i] == '\n') break;
    printf("%x, ", buff[i]);
  }
  printf("\n");
}

void build_buffer(struct Message* msg, unsigned char* buff) {
    buff[0] = msg->start;
    buff[1] = msg->length;
    buff[2] = msg->sequence;
    buff[3] = msg->command_type;
    buff[4] = msg->modality;
    buff[5] = (msg->value >> 8) & 0xFF;
    buff[6] = msg->value & 0xFF;
    msg->crc = calc_crc(buff, msg->length + 5);
    buff[msg->length + 5] = (msg->crc >> 8) & 0xFF;
    buff[msg->length + 6] = msg->crc & 0xFF;
    buff[msg->length + 7] = msg->end;
    print_buffer(buff, sizeof(buff));
}

unsigned short calc_crc(unsigned char* buff, int count) {
    unsigned short crc = 0;
    for (int i = 0; i < count; i++) {
        crc = crc ^ (buff[i] << 8);
        for (int j = 0; j < 8; j++) {
            if (crc & 0x8000) {
                crc = (crc << 1) ^ 0x1021;
            } else {
                crc = crc << 1;
            }
        }
    }
    return crc;
}

